Objects/moduleobject.c

```diff
@@ -784,11 +786,88 @@ _PyModuleSpec_IsUninitializedSubmodule(PyObject *spec, PyObject *name)
     return rc;
 }
 
+static int
+_get_file_origin_from_spec(PyObject *spec, PyObject **origin)
+{
+    PyObject *has_location = NULL;
+    int rc = PyObject_GetOptionalAttr(spec, &_Py_ID(has_location), &has_location);
+    if (rc <= 0) {
+        return rc;
+    }
+    rc = PyObject_IsTrue(has_location);
+    Py_DECREF(has_location);
+    if (rc <= 0) {
+        return rc;
+    }
+    // has_location is true, so origin is a location
+    rc = PyObject_GetOptionalAttr(spec, &_Py_ID(origin), origin);
+    if (rc <= 0) {
+        return rc;
+    }
+    if (!PyUnicode_Check(*origin)) {
+        Py_DECREF(*origin);
+        *origin = NULL;
+        return 0;
+    }
+    return 1;
```

Comment:
Returning a value through an argument can make the function's code a little hard to follow, as a reader.  Consequently, I always try to avoid using the "return" arg variable in the function, using a temporary local variable in its place.  I'd only set the "return" variable right next to the corresponding return statement:

```suggestion
    PyObject *origin = NULL;
    rc = PyObject_GetOptionalAttr(spec, &_Py_ID(origin), &origin);
    if (rc <= 0) {
        return rc;
    }
    if (!PyUnicode_Check(origin)) {
        Py_DECREF(origin);
        return 0;
    }
    *p_origin = origin;
    return 1;
```

That way readers/maintainers don't have to think about the "return" except in one specific place and don't have to distinguish between the two "kinds" of variable.  In my mind, the fewer things someone has to keep track of in the code, the more they can pay attention to the important details and the less likely they are to miss something.

(To be clear, I approve of the approach you took with an `int` return type and returning the object via a pointer argument.  Generally, I'm a fan of keeping the error result separate from the returned object, and in C we don't have multiple value returns, so we don't have other options than returning values through pointer arguments.)